File: App/Resources/AlgorithmDetails/iterativetopdownmergesort/iterativetopdownmergesort.go
package main

import (
	"fmt"
)

func sort(arr []int) []int {
	n := len(arr)
	if n < 2 {
		return arr
	}
	scratch := make([]int, n)
	subarrayCount := 1
	for subarrayCount < n {
		subarrayCount *= 2
	}

	for subarrayCount > 1 {
		for i := 0; i < subarrayCount; i += 2 {
			low := n * i / subarrayCount
			mid := n * (i + 1) / subarrayCount
			high := n * (i + 2) / subarrayCount
			merge(arr, scratch, low, mid, high)
		}
		subarrayCount /= 2
	}
	return arr
}

func merge(arr, scratch []int, low, mid, high int) {
	left, right, out := low, mid, low
	for left < mid && right < high {
		if arr[left] <= arr[right] {
			scratch[out] = arr[left]
			left++
		} else {
			scratch[out] = arr[right]
			right++
		}
		out++
	}
	for left < mid {
		scratch[out] = arr[left]
		left++
		out++
	}
	for right < high {
		scratch[out] = arr[right]
		right++
		out++
	}
	copy(arr[low:high], scratch[low:high])
}

func main() {
	array := []int{0, 39, 21, 62, 91, 77, 14, 23,
		90, 69, 51, 81, 68, 83, 32, 56}
	fmt.Println(sort(array))
}

File: App/Resources/AlgorithmDetails/iterativetopdownmergesort/iterativetopdownmergesort.rb
def sort(arr)
  n = arr.length
  return arr if n < 2
  scratch = Array.new(n)
  subarray_count = 1
  while subarray_count < n
    subarray_count *= 2
  end

  while subarray_count > 1
    i = 0
    while i < subarray_count
      low = n * i / subarray_count
      mid = n * (i + 1) / subarray_count
      high = n * (i + 2) / subarray_count
      merge(arr, scratch, low, mid, high)
      i += 2
    end
    subarray_count /= 2
  end
  arr
end

def merge(array, scratch, low, mid, high)
  left = low
  right = mid
  out = low
  while left < mid && right < high
    if array[left] <= array[right]
      scratch[out] = array[left]
      left += 1
    else
      scratch[out] = array[right]
      right += 1
    end
    out += 1
  end
  while left < mid
    scratch[out] = array[left]
    left += 1
    out += 1
  end
  while right < high
    scratch[out] = array[right]
    right += 1
    out += 1
  end
  (low...high).each { |i| array[i] = scratch[i] }
end


array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
